lib/api/src/store.rs

```diff
@@ -35,10 +34,14 @@ impl Store {
         Self::new_with_tunables(engine, BaseTunables::for_target(engine.target()))
     }
 
-    /// Set the trap handler in this store.
-    pub fn set_trap_handler(&self, handler: Option<Box<TrapHandlerFn>>) {
-        let mut m = self.trap_handler.write().unwrap();
-        *m = handler;
+    /// Sets a [`TrapHandler`] for the `Store`
+    pub fn set_trap_handler(&mut self, handler: Option<Box<TrapHandler<'static>>>) {
+        self.trap_handler = Arc::new(handler);
```

Comment:
So I like this, but lets make sure the new semantics are what we intended. In the old code, calling `set_trap_handler` would globally update the trap handler in all `Store`s. In the new code it only updates the trap handler for this specific store (and Stores `Clone`d from it). All the old trap handlers will continue to work here because they're kept alive by the `Arc`. (This is actually a great example of how using the raw pointer could have caused problems, the moment `set` is called again, the trap handler being used could have been invalid.)

If the move from `Store`'s having 1 trap handler to `Store`s having N trap handlers in a tree-like structure is intentional then 👍 , if not then we probably need to rethink some of this.

And if this is intentional we may want to call it out more in the doc comments.

Reply:
Also side note but if you want to remove the extra `Box` in `Arc<Option<Box<TrapHandler<'static>>>` let me know, I think it should be possible to do so in many cases.

Reply:
I agree with the first comment :-).